Drop support for unknown or unset validate-with

kgaillot opened the pull request.

kgaillot commented:
@nrwahl2 , can you review this one when you get a chance? Thanks

kgaillot commented:
We can't do an XSL tranform for nonconforming CIBs because there's no valid initial schema version to use as the base. Not sure if there would be some way around that, but probably not worth any effort.

nrwahl2 commented:
> We can't do an XSL tranform for nonconforming CIBs because there's no valid initial schema version to use as the base. Not sure if there would be some way around that, but probably not worth any effort.

We already have a transform that adds validate-with="pacemaker-4.0" if it's unset. That can happen because it's an optional attribute per the schema.

We don't handle unknown ones specially. We couldn't (without changes that probably aren't worth the effort, as you said), because the schema we enumerate the valid values, so as you said, there would be no valid initial schema version.

kgaillot commented:
Updated commit to address coverity complaint (we can simplify some code now that original_schema is guaranteed non-NULL), and added an unrelated commit to address another coverity complaint

kgaillot commented:
Rebased on current main to resolve conflict, and dropped last commit (which might re-trigger a coverity complaint, but that commit was moved to a different PR with other coverity fixes)

kgaillot commented:
Separated first commit into multiple for easier review

kgaillot commented:
The only CI failures are CentOS 10 RPMs b/c EPEL 10 is broken

nrwahl2 commented on cts/cts-cli.in:
```diff
@@ -2890,30 +2890,16 @@ class ValidityRegressionTest(RegressionTest):
                  "cibadmin -p -R",
                  stdin=StdinCmd("""cibadmin -Q | sed 's#[ 	][ 	]*validate-with="[^"]*"##'"""),
                  update_cib=True),
-            Test("Run crm_simulate with valid CIB, but without validate-with attribute",
```
Why keep the final `crm_simulate` one ("Run crm_simulate with invalid CIB, also without validate-with attribute")?

kgaillot replied:
It's no longer possible to run crm_simulate on a CIB with a missing or unknown validate-with, so those had to be dropped. Keeping one crm_simulate test with validate-with="none" seemed worthwhile to ensure the nonconforming syntax is actually getting through (as long as we support that).

nrwahl2 reviewed (commented):
Falling asleep, reviewed first four; last two shouldn't take long in the morning

nrwahl2 commented on lib/common/tests/schemas/pcmk__get_schema_test.c:
```diff
@@ -57,9 +58,6 @@ unknown_schema(void **state)
 known_schema(void **state)
 {
-    // @COMPAT none is deprecated since 2.1.8
-    assert_schema(NULL, 15); // defaults to "none"
```
could still test "none" explicitly for now, I don't care

nrwahl2 commented on xml/Makefile.am:
```diff
@@ -210,20 +210,18 @@ pacemaker-%.rng: $(CIB_build_copies) rng-helper Makefile.am
 # Dynamically generated CIB schema listing all pacemaker versions
 #
-# @COMPAT none is deprecated since 2.1.8, as is validate-with being optional
+# @COMPAT "none" is deprecated since 2.1.8
```
Currently, the XSL transform to 4.0 adds `validate-with="pacemaker-4.0"` if `validate-with` is unset. I think it would be straightforward to change that to `validate-with="none"` if desired, to minimize behavioral changes.

That may not be worthwhile, since we plan to drop "none" when we can.

kgaillot replied:
Right, don't want to shift someone to none only to drop it later

nrwahl2 replied:
This change actually breaks the transformation mentioned above, since unset `validate-with` now fails to validate. It's probably better to just live with that.

nrwahl2 commented on lib/cib/cib_file.c:
```diff
@@ -392,9 +392,6 @@ load_file_cib(const char *filename, xmlNode **output)
-
-        crm_err("CIB does not validate against %s, or that schema is unknown", schema);
         pcmk__xml_free(root);
         return -pcmk_err_schema_validation;
```
Should we switch to an earlier `return -pcmk_err_cib_corrupt` in this function too?

kgaillot replied:
I'd rather let `pcmk__configured_schema_validates()` handle it, it already has an error message

nrwahl2 commented on daemons/controld/controld_join_dc.c:
```diff
@@ -539,9 +539,10 @@ do_dc_join_filter_offer(long long action,
+                    "generation) due to %s schema version %s " QB_XS " ref=%s",
+                    join_id, join_from,
+                    ((validation == NULL)? "missing" : "unknown"),
+                    pcmk__s(validation, ""), ref);
```
leaves an extra space character in the message if `validation` is `NULL` , but it only really shows up in the verbose form; not worth changing

kgaillot replied:
right, and it should never really happen anyway

nrwahl2 reviewed (commented):
Very nice! Simplifies things quite a lot.

nrwahl2 reviewed (approved):
You've already addressed my most recent review via comments; just explicitly marking it as approved
